src/tools/cephfs/TableTool.cc

```diff
@@ -0,0 +1,287 @@
+    if (!rank_err.empty()) {
+        derr << "Bad rank '" << rank_str << "'" << dendl;
+        usage();
+    }
+  }
+
+  // RADOS init
+  // ==========
+  r = rados.init_with_context(g_ceph_context);
+  if (r < 0) {
+    derr << "RADOS unavailable, cannot scan filesystem journal" << dendl;
+    return r;
+  }
+
+  dout(4) << "connecting to RADOS..." << dendl;
+  rados.connect();
+ 
+  int const pool_id = mdsmap->get_metadata_pool();
+  dout(4) << "resolving pool " << pool_id << dendl;
+  std::string pool_name;
+  r = rados.pool_reverse_lookup(pool_id, &pool_name);
+  if (r < 0) {
+    derr << "Pool " << pool_id << " named in MDS map not found in RADOS!" << dendl;
+    return r;
+  }
+
+  dout(4) << "creating IoCtx.." << dendl;
+  r = rados.ioctx_create(pool_name.c_str(), io);
+  assert(r == 0);
+
+  // Require 2 args <action> <table>
+  if (argv.size() < 2) {
+    usage();
+    return -EINVAL;
+  }
+  const std::string mode = std::string(argv[0]);
+  const std::string table = std::string(argv[1]);
+
+  if (mode == "reset") {
+    if (table == "session") {
+      return apply_rank_fn(&TableTool::_reset_session_table, NULL);
+    } else if (table == "inode") {
+      return apply_rank_fn(&TableTool::_reset_ino_table, NULL);
+    } else if (table == "snap") {
+      return _reset_snap_table();
+    } else {
+      derr << "Invalid table '" << table << "'" << dendl;
+      usage();
+      return -EINVAL;
+    }
+  } else if (mode == "show") {
+    JSONFormatter jf(true);
+    jf.open_object_section("ranks");
+    if (table == "session") {
+      r = apply_rank_fn(&TableTool::_show_session_table, &jf);
+    } else if (table == "inode") {
+      r = apply_rank_fn(&TableTool::_show_ino_table, &jf);
+    } else if (table == "snap") {
+      r = _show_snap_table(&jf);
+    } else {
+      derr << "Invalid table '" << table << "'" << dendl;
+      usage();
+      return -EINVAL;
+    }
+    jf.close_section();
+
+    // Subcommand should have written to formatter, flush it
+    jf.flush(std::cout);
+    std::cout << std::endl;
+    return r;
+  } else {
+    derr << "Invalid mode '" << mode << "'" << dendl;
+    usage();
+    return -EINVAL;
+  }
+}
+
+
+
+
+
+
+/**
+ * For a function that takes an MDS rank as an argument and
+ * returns an error code, execute it either on all ranks (if
+ * this->rank is MDS_RANK_NONE), or on the rank specified
+ * by this->rank.
+ */
+int TableTool::apply_rank_fn(int (TableTool::*fptr) (mds_rank_t, Formatter*), Formatter *f)
+{
+  int r = 0;
+  if (rank == MDS_RANK_NONE) {
+    std::set<mds_rank_t> in_ranks;
+    mdsmap->get_mds_set(in_ranks);
+
+    for (std::set<mds_rank_t>::iterator rank_i = in_ranks.begin();
+        rank_i != in_ranks.end(); ++rank_i)
+    {
+      if (f) {
+        std::ostringstream rank_str;
+        rank_str << *rank_i;
+        f->open_object_section(rank_str.str().c_str());
+      }
+      int rank_r = (this->*fptr)(*rank_i, f);
+      r = r ? r : rank_r;
```

Comment:
Because we're in disaster recovery, the principle I'm applying is "do as much as you can, even if some stuff is broken".  In this case, because the ranks all use different objects, failures applying a function to one rank should not affect other ranks, so we spit out the error message for the rank that had a problem, and carry on trying to operate on the other ranks.